src/optional/parquet_parser.h

```diff
@@ -0,0 +1,183 @@
+    DMLC_DECLARE_FIELD(format).set_default("parquet")
+      .describe("File format.");
+    DMLC_DECLARE_FIELD(label_column).set_default(0)
+      .describe("Column index (0-based) that will put into label.");
+    DMLC_DECLARE_FIELD(weight_column).set_default(-1)
+      .describe("Column index that will put into instance weights.");
+    DMLC_DECLARE_FIELD(nthreads).set_default(2)
+      .describe("Column index that will put into instance weights.");
+  }
+};
+
+template <typename IndexType, typename DType = real_t>
+class ParquetParser : public ParserImpl<IndexType, DType> {
+
+public:
+  ParquetParser(const std::string& filename,
+                const std::map<std::string, std::string>& args,
+                int nthread) : row_groups_read_(0), nthread_(nthread) {
+    param_.Init(args);
+    nthread_ = param_.nthreads;
+    CHECK_EQ(param_.format, "parquet");
+
+    parquet_reader_ = parquet::ParquetFileReader::OpenFile(filename, false);
+    metadata_ = parquet_reader_->metadata();
+    num_rows_ = metadata_->num_rows();
+    num_cols_ = metadata_->num_columns();
+    num_row_groups_ = metadata_->num_row_groups();
+    
+    have_next_ = (num_rows_ != 0);
+  }
+
+  /*!
+   * \brief read in next several blocks of data
+   * \param data vector of data to be returned
+   * \return true if the data is loaded, false if reach end
+   */
+  virtual bool ParseNext(std::vector<RowBlockContainer<IndexType, DType> > *data);
+
+  protected:
+  virtual void ParseRowGroup(int row_group_id,
+                            RowBlockContainer<IndexType, DType> *out);
+                    
+  virtual size_t BytesRead(void) const {
+    return -1;
+  }
+
+  virtual void BeforeFirst(void) {}
+
+  private:
+    ParquetParserParam param_;
+    // handle for reading parquet files
+    std::unique_ptr<parquet::ParquetFileReader> parquet_reader_;
+    std::shared_ptr<parquet::FileMetaData> metadata_;
+    // number of rows having read
+    int num_rows_;
+    int num_cols_;
+    int num_row_groups_;
+    int row_groups_read_;
+    // whether we have reached end of parquet file
+    bool have_next_;
+    // number of threads; hardcoded 4 for now
+    int nthread_;
+    // OMPException object to catch and rethrow exceptions in omp blocks
+    dmlc::OMPException omp_exc_;
+};
+
+template <typename IndexType, typename DType>
+bool ParquetParser<IndexType, DType>::
+ParseNext(std::vector<RowBlockContainer<IndexType, DType> > *data) {
+  if (!have_next_) {
+    parquet_reader_->Close();
+    return false;
+  }
+  std::vector<std::thread> threads;
+
+  int next_row_groups = std::min(nthread_, num_row_groups_ - row_groups_read_);
+  data->resize(next_row_groups);
+
+  for (int tid = 0; tid < next_row_groups; ++tid) {
+    int row_group_id = row_groups_read_ + tid;
+    threads.push_back(std::thread([this, row_group_id, data, tid] {
+      this->omp_exc_.Run([&] {
+        ParseRowGroup(row_group_id, &(*data)[tid]);
+      });
+    }));
+  }
```

Comment:
Let's use `std::async` instead, so that exceptions are automatically propagated from worker threads to the main threads.

Reply:
Thank you @hcho3 , I've replaced `std::thread` with `std::async`. As the implementation is borrowed from [dmlc-core/src/data/text_parser.h, lines 124-137](https://github.com/dmlc/dmlc-core/blob/main/src/data/text_parser.h#L124-L137), I wonder if that could be updated as well.

Reply:
Let's leave that to a future pull request.